rs/types/types/src/canister_http.rs

```diff
@@ -771,3 +771,187 @@ mod tests {
+        let headers = (0..MAX_CANISTER_HTTP_HEADER_NUM)
+            .map(|i| HttpHeader {
+                name: format!("Header-{}", i),
+                value: "value".to_string(),
+            })
+            .collect::<Vec<_>>();
+        let body = b"";
+
+        let result = validate_http_headers_and_body(&headers, body);
+        assert!(result.is_ok());
+    }
+
+    #[test]
+    fn test_header_name_at_max_length() {
+        // Create a header with name exactly at the limit
+        let headers = vec![HttpHeader {
+            name: "a".repeat(MAX_CANISTER_HTTP_HEADER_NAME_VALUE_LENGTH),
+            value: "value".to_string(),
+        }];
+        let body = b"";
+
+        let result = validate_http_headers_and_body(&headers, body);
+        assert!(result.is_ok());
+    }
+
+    #[test]
+    fn test_header_value_at_max_length() {
+        // Create a header with value exactly at the limit
+        let headers = vec![HttpHeader {
+            name: "Header".to_string(),
+            value: "b".repeat(MAX_CANISTER_HTTP_HEADER_NAME_VALUE_LENGTH),
+        }];
+        let body = b"";
+
+        let result = validate_http_headers_and_body(&headers, body);
+        assert!(result.is_ok());
+    }
+
+    #[test]
+    fn test_headers_at_max_total_size() {
+        // Create headers that sum up exactly to the maximum total size
+        let total_size = MAX_CANISTER_HTTP_HEADER_TOTAL_SIZE;
+
+        let headers = vec![
+            HttpHeader {
+                name: "a".repeat(MAX_CANISTER_HTTP_HEADER_NAME_VALUE_LENGTH),
+                value: "a".repeat(MAX_CANISTER_HTTP_HEADER_NAME_VALUE_LENGTH),
+            },
+            HttpHeader {
+                name: "b".repeat(MAX_CANISTER_HTTP_HEADER_NAME_VALUE_LENGTH),
+                value: "b".repeat(MAX_CANISTER_HTTP_HEADER_NAME_VALUE_LENGTH),
+            },
+            HttpHeader {
+                name: "c".repeat(MAX_CANISTER_HTTP_HEADER_NAME_VALUE_LENGTH),
+                value: "c".repeat(MAX_CANISTER_HTTP_HEADER_NAME_VALUE_LENGTH),
+            },
+        ];
+        let body = b"";
+
+        let result = validate_http_headers_and_body(&headers, body);
+        assert!(result.is_ok());
+    }
+
+    #[test]
+    fn test_too_many_headers() {
+        // Create more headers than allowed
+        let headers = (0..=MAX_CANISTER_HTTP_HEADER_NUM)
+            .map(|i| HttpHeader {
+                name: format!("Header-{}", i),
+                value: "value".to_string(),
+            })
+            .collect::<Vec<_>>();
+        let body = b"";
+
+        let result = validate_http_headers_and_body(&headers, body);
+        assert!(result.is_err());
+        assert!(matches!(
+            result,
+            Err(CanisterHttpRequestContextError::TooManyHeaders(_))
+        ));
+        if let Err(CanisterHttpRequestContextError::TooManyHeaders(count)) = result {
+            assert_eq!(count, MAX_CANISTER_HTTP_HEADER_NUM + 1);
+        }
```

Comment:
I think something like this should work:

```suggestion
        assert!(matches!(
            result,
            Err(CanisterHttpRequestContextError::TooManyHeaders(count)) if count == MAX_CANISTER_HTTP_HEADER_NUM + 1
        ));
```